mypy/join.py

```diff
@@ -234,13 +235,24 @@ def visit_tuple_type(self, t: TupleType) -> Type:
             items = []  # type: List[Type]
             for i in range(t.length()):
                 items.append(self.join(t.items[i], self.s.items[i]))
-            # join fallback types if they are different
             fallback = join_instances(self.s.fallback, t.fallback)
             assert isinstance(fallback, Instance)
             return TupleType(items, fallback)
         else:
             return self.default(self.s)
 
+    def visit_typeddict_type(self, t: TypedDictType) -> Type:
+        if isinstance(self.s, TypedDictType):
+            items = OrderedDict([
+                (item_name, self.join(s_item_type, t_item_type))
+                for (item_name, s_item_type, t_item_type) in self.s.zip(t)
+            ])
+            fallback = join_instances(self.s.fallback, t.fallback)
```

Comment:
It's unclear whether taking the join of the fallbacks is the right thing to do. For example, a join may have fewer items, and the fallback could actually be more specific than operand fallbacks (if it's `Mapping[str, ...]`). Maybe at least add a TODO comment about this.